### ISSUE TITLE
TypeError when performing gauge optimization with respect to fidelity or trace distance

### ISSUE THREAD

**Issue by rileyjmurray** [sandialabs/pyGSTi#406]
Running gauge optimization with respect to fidelity hits the following codeblock:
https://github.com/sandialabs/pyGSTi/blob/0299e1d13222658ab385a17f6e16b9c7ce566163/pygsti/algorithms/gaugeopt.py#L636-L640

The documentation of ``entanglement_fidelity`` says its arguments need to be numpy arrays. However, the code above passes in pyGSTi modelmember objects of one kind or another. If you take a quick look at [the implementation of ``entanglement_fidelity``](https://github.com/sandialabs/pyGSTi/blob/0299e1d13222658ab385a17f6e16b9c7ce566163/pygsti/tools/optools.py#L433-L461) then you'll find that this can easily result in an error. (For example, if either operation is a CompsedOp object.) 

This _specific_ issue could be fixed by replacing 638 -- 640 in the gaugeopt.py snippet to the following:
```python
op_a = target_model.operations[opLbl].to_dense()
op_b = mdl.operations[opLbl].to_dense()
infidelity = 1.0 - _tools.entanglement_fidelity(op_a, op_b, mxBasis)
ret += wt*(infidelity ** 2)
```

That said, _the issue isn't isolated to the snippet above._ Similar bugs exist throughout the definition of the [the objective function for non-LS gauge optimization models](https://github.com/sandialabs/pyGSTi/blob/0299e1d13222658ab385a17f6e16b9c7ce566163/pygsti/algorithms/gaugeopt.py#L594).

**Comment by coreyostrove** [comment]
Great job catching this, @rileyjmurray!

I'm not positive what the history is behind this, but I suspect this is vestigial from the time before more complicated modelmembers for operations existed when pretty much all of the operations inherited from DenseOperatorInterface or similar, and as such behaved in many contexts as if they were numpy arrays. This is clearly no longer true, and your proposed solution makes sense to me.

**Comment by enielse** [comment]
I agree with Corey's assessment - we haven't tried gauge optimizing to
fidelity in a long time and the initial implementation was more
experimental research and there are likely no unit tests.  So yeah,
converting operators to dense matrices first is I think what you want to do.

**Comment by sserita** [comment]
I think this particular bug is basically solved by Timo's PR #414. By performing the `to_dense` cast inside `entanglement_fidelity`, we avoid having to find all the places where the function is called.

The only thing I haven't check as part of that is whether similar bugs exist for other `optool` functions, which I'll do before closing out this issue.

**Comment by sserita** [comment]
Closing as Timo's PR did in fact fix the other possible `optool` functions (`average_gate_fidelity` and `unitarity`).
